docker/android/scripts/build-system.py

```diff
@@ -0,0 +1,207 @@
+    shutil.copy2(src, dst)
+
+
+def restore(dst, args, *_):
+    src = dst + '.bak'
+    if os.path.exists(src):
+        print_verbose(f'restoring from backup "{src}" to "{dst}"', args.verbose)
+        shutil.copy2(src, dst)
+
+
+def filter_map(map, remove):
+    keys = list(map)
+    for key in keys:
+        if not item_op(map[key].value, remove):
+            del map[key]
+    return True
+
+
+def filter_list(lst, remove):
+    lst.filter(lambda x: item_op(x, remove))
+    return True
+
+
+def item_op(item, remove):
+    if item.is_map():
+        return filter_map(item, remove)
+    elif item.is_list():
+        return filter_list(item, remove)
+    elif item.is_string() or item.is_binary_operator():
+        return item.str_op(lambda y: not any(i in y.lower() for i in remove))
+    raise TypeError(f'got unexpected type of {type(item)}')
+
+
+def remove_soong_tests(path, args, *_):
+    print_verbose(f'removing soong tests from "{path}"', args.verbose)
+    with open(path) as file:
+        ast = android.soong.load(file)
+    # remove the test or benchmark scopes, IE, this with `cc_test`
+    # or those with `{name: "test"}`, etc.
+    ast.filter(lambda x: not (x.is_scope() and x.is_dev()))
+    # need to remove test and benchmark subdirs
+    test_names = ('test', 'benchmark')
+    subdirs = [i for i in ast if i.name == 'subdirs']
+    for sub in subdirs:
+        assert type(sub.expr) is android.soong.List
+        filter_list(sub.expr, test_names)
+    # remove gtest dependencies from regular targets.
+    for node in ast:
+        map = None
+        if not node.is_scope() and not node.expr.is_map():
+            continue
+        if node.is_scope():
+            map = node.map
+        else:
+            map = node.expr
+        test_names = ('libgtest', 'test-proto', 'starlarktest')
+        for key, value, *_ in map.recurse():
+            if value.value.is_list():
+                if key == 'testSrcs':
+                    value.value.clear()
+                else:
+                    filter_list(value, test_names)
+
+    with open(path, 'w') as file:
+        ast.dump(file)
+
+
+def remove_makefile_tests(path, args, *_):
+    print_verbose(f'removing makefile tests from "{path}"', args.verbose)
+    with open(path) as file:
+        makefile = android.make.load(file)
+    makefile.filter(lambda x: not x.is_dev())
+    with open(path, 'w') as file:
+        makefile.dump(file)
+
+
+def remove_tests(path, args, processor):
+    if os.path.exists(path + '.bak'):
+        restore(path, args)
+    elif not args.disable_backup:
+        backup(path, args)
+    processor(path, args)
+
+
+def stash(root):
+    git_glob = f'{root}/**/.git'
+    for path in glob.iglob(git_glob, recursive=True):
+        os.chdir(os.path.dirname(path))
+        subprocess.check_call(['git', 'stash'])
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    action_group = parser.add_mutually_exclusive_group(required=True)
+    action_group.add_argument(
```

Comment:
These options are quite handy when debugging: we can use `--backup` to restore from build-system only files, `--stash` allows us to simply stash all changes and start from scratch (rather than have to use `git stash` in every repository ourselves, and `--remove-tests` is our main, desired functionality.